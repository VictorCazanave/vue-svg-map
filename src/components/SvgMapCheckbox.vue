<script setup lang="ts">
import type { SVGAttributes } from 'vue'
import type { Map, Location } from '../types'
import SvgMap from './SvgMap.vue'

defineProps<{map: Map}>()

// Ids of selected locations
const model = defineModel<string[]>({ default: () => [] })

/**
 * Indicate whether a location is selected
 *
 * @param location - Location DOM object
 * @returns True if the location is selected
 */
function isLocationSelected(location: Location): boolean {
  return model.value.some(locationId => locationId === location.id)
}

function getLocationAttributes(location: Location): SVGAttributes {
  return {
    role: 'checkbox',
    tabindex: '0',
    'aria-checked': isLocationSelected(location),
  }
}

/**
 * Select/deselect a location
 *
 * @param event - Triggered event
 */
function toggleLocation(event: MouseEvent | KeyboardEvent): void {
  const target = event.target as SVGElement

  if(target) {
    const locationId = target.id;
    const selectedLocationIndex = model.value.indexOf(locationId)

    // NOTE: Mutating model with splice or push does not emit event (https://github.com/vuejs/core/issues/11143#issuecomment-2173390531)
    if (selectedLocationIndex > -1) {
      // Remove location
      model.value = [...model.value.slice(0, selectedLocationIndex), ...model.value.slice(selectedLocationIndex + 1)];
    } else {
      // Add location
      model.value = [...model.value, locationId];
    }
  }
}
</script>

<template>
  <!-- NOTE: Use .stop modifier to prevent event bubbling and calling function twice -->
	<SvgMap
		v-bind="$attrs"
    :map="map"
		role="group"
    :locationAttributes="getLocationAttributes"
		@click.stop="toggleLocation"
		@keydown.prevent.stop.space="toggleLocation"
	>
		<!-- Pass down slots to SvgMap: https://stackoverflow.com/a/50892881/9826498 -->
		<template
			v-for="(_, slotName) of $slots"
			v-slot:[slotName]="scope"
		>
			<slot
				v-bind="scope"
				:name="slotName"
			/>
		</template>
	</SvgMap>
</template>
