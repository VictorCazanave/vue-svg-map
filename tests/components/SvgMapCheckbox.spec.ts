import { mount } from '@vue/test-utils'
import type { VueWrapper, DOMWrapper } from '@vue/test-utils'
import map from './mocks/map'
import SvgMapCheckbox from '../../src/components/SvgMapCheckbox.vue'

describe('SvgMapCheckbox component', () => {
	const modelValue = ['id0']
	let wrapper: VueWrapper
	let locationElt0: DOMWrapper<SVGPathElement>
	let locationElt1: DOMWrapper<SVGPathElement>
	let locationElt2: DOMWrapper<SVGPathElement>

	beforeEach(() => {
		wrapper = mount(SvgMapCheckbox, { props: { map, modelValue } })

		const locationElts: DOMWrapper<SVGPathElement>[] = wrapper.findAll('path')
    locationElt0 = locationElts.at(0)!
    locationElt1 = locationElts.at(1)!
    locationElt2 = locationElts.at(2)!
	})

  it('renders svg map as a group of checkboxes', () => {
		expect(wrapper.element).toMatchSnapshot()
	})

	it('selects locations of inital model value', () => {
		expect(locationElt0.attributes('aria-checked')).toBe('true')
		expect(locationElt1.attributes('aria-checked')).toBe('false')
		expect(locationElt2.attributes('aria-checked')).toBe('false')
	})

	it('selects locations of new model value', async () => {
		await wrapper.setProps({ modelValue: ['id0', 'id1'] })

		expect(locationElt0.attributes('aria-checked')).toBe('true')
		expect(locationElt1.attributes('aria-checked')).toBe('true')
		expect(locationElt2.attributes('aria-checked')).toBe('false')
	})

	it('selects new location when clicking on an unselected location', async () => {
		await locationElt2.trigger('click')

    expect(locationElt0.attributes('aria-checked')).toBe('true')
		expect(locationElt1.attributes('aria-checked')).toBe('false')
		expect(locationElt2.attributes('aria-checked')).toBe('true')
	})

	it('selects new location when pressing spacebar on an unselected location', async () => {
		await locationElt2.trigger('keydown.space')

    expect(locationElt0.attributes('aria-checked')).toBe('true')
		expect(locationElt1.attributes('aria-checked')).toBe('false')
		expect(locationElt2.attributes('aria-checked')).toBe('true')
	})

	it('unselects location when clicking on a selected location', async () => {
		await locationElt0.trigger('click')

    expect(locationElt0.attributes('aria-checked')).toBe('false')
		expect(locationElt1.attributes('aria-checked')).toBe('false')
		expect(locationElt2.attributes('aria-checked')).toBe('false')
	})

	it('unselects location when pressing spacebar on a selected location', async () => {
		await locationElt0.trigger('keydown.space')

    expect(locationElt0.attributes('aria-checked')).toBe('false')
		expect(locationElt1.attributes('aria-checked')).toBe('false')
		expect(locationElt2.attributes('aria-checked')).toBe('false')
	})

	it('emits update:modelValue event with new selected locations when clicking on an unselected location', async () => {
		await locationElt2.trigger('click')

		expect(wrapper.emitted('update:modelValue')).toEqual([[['id0', 'id2']]])
	})

	it('emits update:modelValue event with new selected locations when pressing spacebar on an unselected location', async () => {
		await locationElt2.trigger('keydown.space')

		expect(wrapper.emitted('update:modelValue')).toEqual([[['id0', 'id2']]])
	})

	it('emits update:modelValue event with new selected locations when clicking on a selected location', async () => {
		await wrapper.setProps({ modelValue: ['id0', 'id1'] })
		await locationElt0.trigger('click')

		expect(wrapper.emitted('update:modelValue')).toEqual([[['id1']]])
	})

	it('emits update:modelValue event with new selected locations when pressing spacebar on a selected location', async () => {
		await wrapper.setProps({ modelValue: ['id0', 'id1'] })
		await locationElt0.trigger('keydown.space')

		expect(wrapper.emitted('update:modelValue')).toEqual([[['id1']]])
	})

	it('emits update:modelValue event with empty array when unselecting all locations', async () => {
		await locationElt0.trigger('click')

		expect(wrapper.emitted('update:modelValue')).toEqual([[[]]])
	})
})
