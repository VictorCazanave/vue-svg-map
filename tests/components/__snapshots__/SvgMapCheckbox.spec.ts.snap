// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SvgMapCheckbox component > renders svg map as a group of checkboxes 1`] = `
<div
  data-v-app=""
>
  
  <!-- NOTE: Use .stop modifier to prevent event bubbling and calling function twice -->
  <svg
    aria-label="label"
    class="svg-map"
    role="group"
    viewBox="viewBox"
    xmlns="http://www.w3.org/2000/svg"
  >
    
    
    
    <path
      aria-checked="true"
      aria-label="name0"
      class="svg-map__location"
      d="path0"
      id="id0"
      role="checkbox"
      tabindex="0"
    />
    <path
      aria-checked="false"
      aria-label="name1"
      class="svg-map__location"
      d="path1"
      id="id1"
      role="checkbox"
      tabindex="0"
    />
    <path
      aria-checked="false"
      aria-label="name2"
      class="svg-map__location"
      d="path2"
      id="id2"
      role="checkbox"
      tabindex="0"
    />
    
    
    
  </svg>
  
</div>
`;
